File: Sample/2.0/Controllers/Sample/ProductsController.cs
using System;
using System.Linq;
using System.Collections.Generic;
using Newtonsoft.Json.Linq;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNet.OData;
using Microsoft.AspNet.OData.Routing;
using Microsoft.EntityFrameworkCore;
using Microsoft.EntityFrameworkCore.Metadata.Internal;

namespace MyApp.Controllers.Sample
{
  using Models;
  using Data;
  using Models.Sample;

  [EnableQuery]
  [ODataRoutePrefix("odata/Sample/Products")]
  public partial class ProductsController : Controller
  {
    private Data.SampleContext context;

    public ProductsController(Data.SampleContext context)
    {
      this.context = context;
    }
    // GET /odata/Sample/Products
    [HttpGet]
    public IEnumerable<Models.Sample.Product> GetProducts()
    {
      var items = this.context.Products.AsQueryable<Models.Sample.Product>();

      this.OnProductsRead(ref items);

      return items;
    }

    partial void OnProductsRead(ref IQueryable<Models.Sample.Product> items);

    [HttpGet("{Id}")]
    public IActionResult GetProduct(int key)
    {
        var item = this.context.Products.Where(i=>i.Id == key).SingleOrDefault();

        if (item == null)
        {
            return NotFound();
        }

        return new ObjectResult(item);
    }
    partial void OnProductDeleted(Models.Sample.Product item);

    [HttpDelete("{Id}")]
    public IActionResult DeleteProduct(int key)
    {
        var item = this.context.Products
            .Where(i => i.Id == key)
            .Include(i => i.OrderDetails)
            .SingleOrDefault();

        if (item == null)
        {
            return NotFound();
        }

        this.OnProductDeleted(item);
        this.context.Products.Remove(item);
        this.context.SaveChanges();

        return new NoContentResult();
    }

    partial void OnProductUpdated(Models.Sample.Product item);

    [HttpPut("{Id}")]
    public IActionResult PutProduct(int key, [FromBody]Models.Sample.Product newItem)
    {
        if (newItem == null || newItem.Id != key)
        {
            return BadRequest();
        }

        this.OnProductUpdated(newItem);
        this.context.Products.Update(newItem);
        this.context.SaveChanges();

        return new NoContentResult();
    }

    [HttpPatch("{Id}")]
    public IActionResult PatchProduct(int key, [FromBody]JObject patch)
    {
        var item = this.context.Products.Where(i=>i.Id == key).FirstOrDefault();

        if (item == null)
        {
            return BadRequest();
        }

        Data.EntityPatch.Apply(item, patch);

        this.OnProductUpdated(item);
        this.context.Products.Update(item);
        this.context.SaveChanges();

        return new NoContentResult();
    }

    partial void OnProductCreated(Models.Sample.Product item);

    [HttpPost]
    public IActionResult Post([FromBody] Models.Sample.Product item)
    {
        if (item == null)
        {
            return BadRequest();
        }

        this.OnProductCreated(item);
        this.context.Products.Add(item);
        this.context.SaveChanges();

        return Created($"odata/Sample/Products/{item.Id}", item);
    }
  }
}

File: Sample/2.0_VB/Controllers/Sample/ProductsController.vb
Imports Newtonsoft.Json.Linq
Imports Microsoft.AspNetCore.Mvc
Imports Microsoft.AspNet.OData
Imports Microsoft.AspNet.OData.Routing
Imports Microsoft.EntityFrameworkCore

Namespace MyApp.Controllers.Sample

    <EnableQuery>
    <ODataRoutePrefix("odata/Sample/Products")>
    Partial Public Class ProductsController
        Inherits Controller
        Private context As Data.SampleContext

        Public Sub New(context As Data.SampleContext)
            Me.context = context
        End Sub

        <HttpGet>
        Public Function GetProducts() As IEnumerable(Of Models.Sample.Product)
            Dim items = Me.context.Products.AsQueryable()

            Me.OnProductsRead(items)

            Return items
        End Function

        Partial Private Sub OnProductsRead(ByRef items As IQueryable(Of Models.Sample.Product))
        End Sub

        <HttpGet("{Id}")>
        Public Function GetProduct(key As Integer) As IActionResult
            Dim item = Me.context.Products.Where(Function(i) i.Id = key).SingleOrDefault()

            If item Is Nothing Then
                Return NotFound()
            End If

            Return New ObjectResult(item)
        End Function
        Partial Private Sub OnProductDeleted(item As Models.Sample.Product)
        End Sub

        <HttpDelete("{Id}")>
        Public Function DeleteProduct(key As Integer) As IActionResult
            Dim item = Me.context.Products.Where(Function(i) i.Id = key).Include(Function(i) i.OrderDetails).SingleOrDefault()

            If item Is Nothing Then
                Return NotFound()
            End If

            Me.OnProductDeleted(item)
            Me.context.Products.Remove(item)
            Me.context.SaveChanges()

            Return New NoContentResult()
        End Function

        Partial Private Sub OnProductUpdated(item As Models.Sample.Product)
        End Sub

        <HttpPut("{Id}")>
        Public Function PutProduct(key As Integer, <FromBody> newItem As Models.Sample.Product) As IActionResult
            If newItem Is Nothing OrElse newItem.Id <> key Then
                Return BadRequest()
            End If

            Me.OnProductUpdated(newItem)
            Me.context.Products.Update(newItem)
            Me.context.SaveChanges()

            Return New NoContentResult()
        End Function

        <HttpPatch("{Id}")>
        Public Function PatchProduct(key As Integer, <FromBody> patch As JObject) As IActionResult
            Dim item = Me.context.Products.Where(Function(i) i.Id = key).FirstOrDefault()

            If item Is Nothing Then
                Return BadRequest()
            End If

            Data.EntityPatch.Apply(item, patch)

            Me.OnProductUpdated(item)
            Me.context.Products.Update(item)
            Me.context.SaveChanges()

            Return New NoContentResult()
        End Function

        Partial Private Sub OnProductCreated(item As Models.Sample.Product)
        End Sub

        <HttpPost>
        Public Function Post(<FromBody> item As Models.Sample.Product) As IActionResult
            If item Is Nothing Then
                Return BadRequest()
            End If

            Me.OnProductCreated(item)
            Me.context.Products.Add(item)
            Me.context.SaveChanges()

            Return Created("odata/Sample/Products/{item.Id}", item)
        End Function

    End Class

End Namespace
